### ISSUE TITLE
Regression in TypeScript 5.4: "Excessive stack depth comparing types" errors

### ISSUE THREAD

**Issue by MajorLift** [poteat/hkt-toolbelt#80]
An unknown breaking change in TypeScript v5.4 is causing errors:
- The first stable release that triggers this issue is `v5.4.2`
- TypeScript version downgraded to `v5.3.3` in main branch as a temporary fix: 8843a718939021ff256afe25317169a2675ff84b
- All 47 errors are "TS2321: Excessive stack depth comparing types..."

```ts
src/natural-number/modulo.ts:15:46 - error TS2321: Excessive stack depth comparing types '_$divide<A_LIST, B_LIST, "MODULO">' and 'readonly Digit[]'.

15   MODULUS_LIST extends DigitList.DigitList = DigitList._$modulo<A_LIST, B_LIST>,
                                                ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Found 47 errors in 10 files.

Errors  Files
     3  src/digit-list/divide.ts:70
     8  src/integer/divide-by.ts:7
     8  src/integer/divide.ts:61
     1  src/integer/modulo.ts:31
     1  src/integer/remainder.ts:22
     8  src/list/range.ts:20
     1  src/list/reverse.ts:30
     8  src/natural-number/divide-by.ts:7
     8  src/natural-number/divide.ts:47
     1  src/natural-number/modulo.ts:15
```

> CI run error log: https://github.com/poteat/hkt-toolbelt/actions/runs/8399043832/job/23004659944

- These errors were previously masked in the CI runs by a failing linter workflow.
  - First failing CI run: https://github.com/poteat/hkt-toolbelt/actions/runs/8309157179/job/22740200012
  - Linter build scripts fixed here (cause is unrelated to this issue): https://github.com/poteat/hkt-toolbelt/commit/9160d65dbfc9b931b38f3b16fc89e17eac1e9b3f

**Comment by MajorLift** [comment]
- Possible cause: https://github.com/microsoft/TypeScript/pull/56004
- Possible solution: https://github.com/sindresorhus/type-fest/pull/807

**Comment by poteat** [comment]
Fixed by #85
